### PR TITLE
feat: add Preceder Metadata Frame for per-object streaming metadata

### PR DESCRIPTION
**Pull request by tlmquintino**
## Summary

- Add **Preceder Metadata Frame** (frame type 8) — a new optional frame that immediately precedes a DataObject frame, carrying per-object metadata for streaming producers that may not know ahead of time when the message will end.
- CBOR structure reuses `GlobalMetadata` with `common` empty and a single-entry `payload` array. Key-level merge on decode: preceder keys override footer keys, structural keys from footer preserved.
- New API: `StreamingEncoder::write_preceder()`, C FFI `tgm_streaming_encoder_write_preceder()`, C++ wrapper `streaming_encoder::write_preceder()`.

## Wire Format Changes

| Item | Value |
|------|-------|
| Frame type | `PrecederMetadata = 8` |
| Preamble flag | Bit 6 (`PRECEDER_METADATA`) |
| Ordering | Must immediately precede one DataObject, same phase as DataObjects |
| CBOR structure | GlobalMetadata with empty common, single-entry payload array |
| Merge rule | Preceder keys win over footer payload[i] keys |

## Test Coverage

- 279 Rust tests (29 preceder-specific across wire, framing, streaming, edge_cases)
- 109 C++ tests (4 new preceder tests)
- 97 Python tests (unchanged — decode-side is transparent)
- 0 clippy warnings, formatting clean

## Files Changed (19 files, +1146 / -22 lines)

**Core Rust:** wire.rs, framing.rs, streaming.rs, encode.rs, iter.rs
**Bindings:** tensogram-ffi (lib.rs + tensogram.h), tensogram.hpp, tensogram-python
**Tests:** edge_cases.rs, golden_files.rs, test_streaming.cpp
**Docs:** WIRE_FORMAT.md, DONE.md, wire-format.md, metadata.md, edge-cases.md

Docs Preview
https://sites.ecmwf.int/docs/dev-section/tensogram/pull-requests/PR-8

### CODE REVIEW

File: plans/WIRE_FORMAT.md [hunk 1]
```diff
@@ -53,8 +54,11 @@ Here is the overall message format:
 ├──────────────────────────────────────────────────────────────┤
 │  HEADER HASH FRAME CBOR [count, hash_type, hashes] (optional)│
 ├──────────────────────────────────────────────────────────────┤
+│  PRECEDER METADATA FRAME (for object 0, optional)            │
 │  DATA OBJECT FRAME  header + payload + footer  (object 0)    │
+│  PRECEDER METADATA FRAME (for object 1, optional)            │
 │  DATA OBJECT FRAME  header + payload + footer  (object 1)    │
+���  DATA OBJECT FRAME  header + payload + footer  (object 2)    │
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply plans/WIRE_FORMAT.md]: fix

File: tests/cpp/test_streaming.cpp [hunk 2]
```diff
@@ -221,3 +221,99 @@ TEST(StreamingTest, StreamingWithHash) {
...
+            values.size() * sizeof(float));
+
+        enc.finish();
+    }
+
+    // Open and decode — preceder metadata should be in payload[0]
+    auto f = tensogram::file::open(tmp.path);
+    auto msg = f.decode_message(0);
+    EXPECT_EQ(msg.num_objects(), 1u);
+
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply tests/cpp/test_streaming.cpp]: fix the comment

File: crates/tensogram-core/src/framing.rs [hunk 3]
```diff
@@ -577,24 +600,64 @@ pub fn decode_message(buf: &[u8]) -> Result<DecodedMessage<'_>> {
...
     })?;
 
+    // Merge preceder payloads into global_metadata.payload (preceder wins).
+    // Key-level merge: preceder keys override footer keys on conflict,
+    // but footer-only keys (e.g. auto-populated ndim/shape/strides/dtype)
+    // are preserved when absent from the preceder.
+    let obj_count = objects.len();
+    global_metadata
+        .payload
+        .resize_with(obj_count, BTreeMap::new);
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply crates/tensogram-core/src/framing.rs]: adopt suggestions

File: plans/WIRE_FORMAT.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (tlmquintino) [reply plans/WIRE_FORMAT.md]: Fixed in 3b35b69 — replaced corrupted character with correct box-drawing `│`.

File: tests/cpp/test_streaming.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (tlmquintino) [reply tests/cpp/test_streaming.cpp]: Fixed in 3b35b69 — updated comment to accurately describe what the test verifies (data round-trip, not metadata keys), and added per-element float assertions.

File: crates/tensogram-core/src/framing.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tlmquintino) [reply crates/tensogram-core/src/framing.rs]: Adopted in 3b35b69 — replaced `resize_with` with validate+extend: returns `Metadata` error when `payload.len() > obj_count`, only extends when shorter. Added `test_decode_rejects_payload_count_exceeding_objects` in d637e74 to cover this path.
